stock_dynamic_routing/models/stock_move.py

```diff
@@ -318,64 +351,55 @@ def _apply_routing_rule_pull(self, routing_details):
+        new move but switch the source location of the current move.  This
+        might trigger a new routing on the destination move.
         """
-        self.location_dest_id = routing_rule.location_dest_id
-        next_move = self.move_dest_ids.filtered(lambda r: r.state == "waiting")
-        # FIXME what should happen if we have > 1 move? What would
-        # be the use case?
-        if next_move and len(next_move) == 1:
-            split_move = next_move
-            split_move_vals = next_move._split(self.product_qty)
-            if split_move_vals:
-                split_move = self.create(split_move_vals)
-                split_move._action_confirm(merge=False)
-            if split_move != next_move:
-                # No split occurs if the quantity was the same.
-                # But if it did split, detach it from the move on which
-                # we have a different routing
-                next_move.move_orig_ids -= self
-                split_move.move_orig_ids = self
-            next_move = split_move
-
-        next_move.location_id = routing_rule.location_dest_id
-
-    def _routing_pull_insert_move(self, routing_rule, picking_type, destination):
-        """Add a move after the current move to reach the destination
-
-        The routing rules are applied on the new move in case it would trigger
-        a new move or a switch of picking type.
-        """
-        self.location_dest_id = routing_rule.location_dest_id
-        # create a copy of the move with the current picking type and
-        # going to its original destination: it will be assigned to the
-        # same picking as the original picking of our move
-        routing_move = self._insert_routing_moves(
-            picking_type,
-            # the source of the next move has to be the same as the
-            # destination of the current move
-            routing_rule.location_dest_id,
-            destination,
-        )
-        routing_move._chain_apply_routing()
+        for move in self:
+            origmoves_by_location = OrderedDict()
+            for orig_move in move.move_orig_ids:
+                origmoves_by_location.setdefault(
+                    orig_move.location_dest_id, move.browse()
+                )
+                origmoves_by_location[orig_move.location_dest_id] |= orig_move
+            for location_id, orig_moves in reversed(origmoves_by_location.items()):
+                qty = sum(orig_moves.mapped("product_qty"))
+                split_move = move
+                split_move_vals = move._split(qty)
+                if split_move_vals:
+                    split_move = self.create(split_move_vals)
+                if split_move != move:
+                    # No split occurs if the quantity was the same.
+                    # But if it did split, detach it from the move on which
+                    # we have a different routing
+                    move.move_orig_ids -= orig_moves
```

Comment:
minor remark: it's ok for perfs to make a write in a for ?

Reply:
We didn't spotted any issue regarding performance so far, but yes that could be improved maybe especially the use `|= orig_move` above to instead use IDs.
=> https://gist.github.com/guewen/a4a2eba5d128b4ed3129866f80571fbd